### PR TITLE
ELMU #180 Make hard-deletions optionally delete all revisions

### LINKED ISSUE #180: Deleting a plugin in both directions (forward and backward) in the history of a document
If an image plugin needs to be hard deleted for copyright reasons, you can only delete subsequent occurrences, but not preceding ones. An option would be helpful to delete all occurrences of an plugin (for example image plugin) in the history of a document.

### PR DESCRIPTION
**Pull request by ahelmberger**
Also ensure to not double-hard-delete
Also fix some glitches on the hard-delete dialog UI

Closes #180

### CODE REVIEW

File: src/components/confirmation-dialogs.js [hunk 1]
```diff
@@ -19,24 +19,23 @@ export function confirmSectionDelete(t, section, onOk, onCancel = () => {}) {
...
 ) {
   let dialog = null;
-  let deletionReason = '';
-  let deleteDescendants = false;
-  let handleDeletionReasonChange = null;
-  let handleDeleteDescendantsChange = null;
+  let reason = '';
+  let deleteAllRevisions = false;
+  let handleReasonChange = null;
+  let handleDeleteAllRevisionsChange = null;
```
Comment (adrianaluchian) [inline src/components/confirmation-dialogs.js:29]: Small thing, we could inline these variables, I think there's no need for them to be instantiated with null.

File: src/components/pages/docs.js [hunk 2]
```diff
@@ -137,6 +137,16 @@ class Docs extends React.Component {
...
 
+  renderSlug(value, doc) {
+    if (!doc.slug) {
+      const { t } = this.props;
+      return t('notAssigned');
+    }
+
+    const url = urls.getArticleUrl(doc.slug);
+    return <a href={url}>{url}</a>;
+  }
```
Comment (adrianaluchian) [inline src/components/pages/docs.js:148]: `value` doesn't seem used here (anymore).

File: src/components/pages/docs.js [hunk 3]
```diff
@@ -227,14 +244,14 @@ class Docs extends React.Component {
...
-            <p>{t('title')}</p>
-            <p><Input value={newDocTitle} onChange={this.handleNewDocTitleChange} /></p>
-            <p>{t('language')}</p>
-            <p><LanguageSelect value={newDocLanguage} onChange={this.handleNewDocLanguageChange} /></p>
-            <p>{t('slug')}</p>
-            <p><Input addonBefore={urls.articlesPrefix} value={newDocSlug} onChange={this.handleNewDocSlugChange} /></p>
-            {isLoading && <p>{t('newDocumentProgress')}</p>}
+            <div>{t('title')}</div>
+            <div><Input value={newDocTitle} onChange={this.handleNewDocTitleChange} /></div>
+            <div style={{ marginTop: '0.75em' }}>{t('language')}</div>
```
Comment (adrianaluchian) [inline src/components/pages/docs.js:251]: Could we put this in `docs.less` maybe?

File: src/services/document-service.spec.js [hunk 4]
```diff
@@ -114,4 +114,191 @@ describe('document-service', () => {
...
+            slug: 'rev-5',
+            sections: [
+              { ...cloneDeep(unrelatedSection), content: { text: 'Unrelated E' } },
+              { ...cloneDeep(sectionToBeDeleted), content: { text: 'Doomed section C' } }
+            ]
+          }
+        ]);
+      });
+
+      describe('and only the second revision is hard-deleted', () => {
```
Comment (adrianaluchian) [inline src/services/document-service.spec.js:183]: A little confusing here with `the second revision` and `documentRevisionsBeforeDeletion[2]`

File: src/services/document-service.spec.js [hunk 5]
```diff
@@ -114,4 +114,191 @@ describe('document-service', () => {
...
+            deletedBecause: null,
+            content: { text: 'Doomed section A' }
+          });
+          expect(documentRevisionsAfterDeletion[4].sections[1]).toMatchObject({
+            deletedOn: null,
+            deletedBy: null,
+            deletedBecause: null,
+            content: { text: 'Doomed section C' }
+          });
+        });
```
Comment (adrianaluchian) [inline src/services/document-service.spec.js:233]: Maybe we could add a lighter test that checks that none of the other sections are touched at all (not property by property but deep equal)

File: src/services/document-service.spec.js [hunk 6]
```diff
@@ -114,4 +114,191 @@ describe('document-service', () => {
...
+        it('deletes all occurrences of that section', () => {
+          documentRevisionsAfterDeletion.forEach(revision => {
+            expect(revision.sections[1]).toMatchObject({
+              deletedOn: expect.any(Date),
+              deletedBy: user._id,
+              deletedBecause: 'My reason',
+              content: null
+            });
+          });
+        });
```
Comment (adrianaluchian) [inline src/services/document-service.spec.js:261]: Same here, a test like `keeps all other sections`.

File: src/services/document-service.spec.js (hunk 4 above)
Comment (ahelmberger) [reply src/services/document-service.spec.js:183]: True, maybe `the revision in the middle`?

File: src/components/pages/docs.js (hunk 2 above)
Comment (adrianaluchian) [reply src/components/pages/docs.js:148]: Got it!

File: src/components/confirmation-dialogs.js (hunk 1 above)
Comment (ahelmberger) [reply src/components/confirmation-dialogs.js:29]: Do you mean to not initialize them with default values? Because to inline would mean to me moving their declaration to where they're used (which is not possible as they're used in multiple places).

**Review by laurentiu-ilici — APPROVED** [review]

File: src/components/pages/docs.js [hunk 7]
```diff
@@ -137,6 +137,16 @@ class Docs extends React.Component {
...
   }
 
+  renderSlug(value, doc) {
+    if (!doc.slug) {
+      const { t } = this.props;
+      return t('notAssigned');
+    }
+
+    const url = urls.getArticleUrl(doc.slug);
+    return <a href={url}>{url}</a>;
```
[the remark this answers is not in the document]
Comment (ahelmberger) [reply src/components/pages/docs.js:147]: The name (=title) is already in the first column of the table, this column has the header `URL`, so we should really show the URL, I think.

File: src/components/pages/docs.js (hunk 3 above)
Comment (ahelmberger) [reply src/components/pages/docs.js]: Done

File: src/services/document-service.spec.js (hunk 5 above)
Comment (ahelmberger) [reply src/services/document-service.spec.js:233]: Done
